uniffi_macros/src/record.rs

```diff
@@ -0,0 +1,96 @@
+use proc_macro2::{Span, TokenStream};
+use quote::quote;
+use syn::{Data, DeriveInput};
+use uniffi_meta::{FieldMetadata, RecordMetadata};
+
+use crate::{export::metadata::convert::convert_type, util::create_metadata_static_var};
+
+pub fn expand_record(input: DeriveInput, module_path: Vec<String>) -> TokenStream {
+    let fields = match input.data {
+        Data::Struct(s) => Some(s.fields),
+        _ => None,
+    };
+
+    let ident = &input.ident;
+
+    let (write_impl, try_read_fields) = match &fields {
+        Some(fields) => fields
+            .iter()
+            .map(|f| {
+                let ident = &f.ident;
+                let ty = &f.ty;
+
+                let write_field = quote! {
+                    <#ty as ::uniffi::FfiConverter>::write(obj.#ident, buf);
+                };
+                let try_read_field = quote! {
+                    #ident: <#ty as ::uniffi::FfiConverter>::try_read(buf)?,
+                };
+
+                (write_field, try_read_field)
+            })
+            .unzip(),
+        None => {
+            let unimplemented = quote! { ::std::unimplemented!() };
+            (unimplemented.clone(), unimplemented)
```

Comment:
I think this could actually be implemented with an empty body or maybe a `()` statement.  I don't think we have much use for unit structs as records, but it would there be a downside to implementing them?

Reply:
I don't know whether there is a downside to allowing this derive on unit structs, but note that this code path is also taken when the input is an `enum` or `union`. If it is, we also emit a compile error (line 71).

Reply:
Ahh, that makes a lot of sense.